py2pack/__init__.py

```diff
@@ -311,29 +346,41 @@ def generate(args):
     print('generating spec file for {0}...'.format(args.name))
     data = args.fetched_data['info']
     durl = newest_download_url(args)
-    data['source_url'] = (args.source_url or
-                          (durl and durl['url']) or
-                          args.name + '-' + args.version + '.zip')
+    source_url = data['source_url'] = (args.source_url or (durl and durl['url']))
     data['year'] = datetime.datetime.now().year                             # set current year
     data['user_name'] = pwd.getpwuid(os.getuid())[4]                        # set system user (packager)
     data['summary_no_ending_dot'] = re.sub(r'(.*)\.', r'\g<1>', data.get('summary', ""))
 
     # If package name supplied on command line differs in case from PyPI's one
     # then package archive will be fetched but the name will be the one from PyPI.
     # Eg. send2trash vs Send2Trash. Check that.
-    for name in (args.name, data['name']):
-        tarball_file = glob.glob("{0}-{1}.*".format(name, args.version))
-        # also check tarball files with underscore. Some packages have a name with
-        # a '-' or '.' but the tarball name has a '_' . Eg the package os-faults
-        tr = str.maketrans('-.', '__')
-        tarball_file += glob.glob("{0}-{1}.*".format(name.translate(tr),
-                                                     args.version))
+    tr = str.maketrans('-.', '__')
+    version = args.version
+    name = args.name
+    try:
+        source_glob = args.source_glob
+    except AttributeError:
+        source_glob = '%{name}-%{version}.*'
+    data_name = data['name'] or name
+
+    tarball_file = []
+    for __name in (name, name.translate(tr), data_name, data_name.translate(tr)):
+        tarball_file.extend(glob.glob(replace_string(source_glob, {'name': __name, 'version': version})))
+        if tarball_file:
+            break
```

Comment:
Why do you use this new function `replace_string` instead of just keeping the python format for this simple replacement?

```suggestion
        source_glob = '{name}-{version}.*'
    data_name = data['name'] or name

    tarball_file = []
    for n in (name, name.translate(tr), data_name, data_name.translate(tr)):
        tarball_file.extend(glob.glob(source_glob.format(name=n, version=version)))
        if tarball_file:
            break
```

Reply:
Because Python's 'format' method for replacement explicitly replaces all elements with {name}, but sometimes we need to leave {name} unchanged, also Python's 'format' method does not look for either % symbol, or %% symbol.

Reply:
I don't get why it's required to be able to look for `{name}` or something like that in the glob search for filenames, but if that is your user case, it's okay, not a big deal.